neuro_san/internals/run_context/langchain/llms/azure_client_policy.py

```diff
@@ -0,0 +1,98 @@
+#
+# END COPYRIGHT
+from typing import Any
+from typing import Dict
+
+from leaf_common.config.resolver import Resolver
+
+from neuro_san.internals.run_context.langchain.llms.openai_client_policy import OpenAIClientPolicy
+
+
+class AzureClientPolicy(OpenAIClientPolicy):
+    """
+    ClientPolicy implementation for OpenAI via Azure.
+
+    OpenAI's BaseLanguageModel implementations do allow us to pass in a web client
+    as an argument, so this implementation takes advantage of the create_client()
+    method to do that. Worth noting that where many other implementations might care about
+    the llm reference, because of our create_client() implementation, we do not.
+    """
+
+    def create_client(self, config: Dict[str, Any]) -> Any:
+        """
+        Creates the web client to used by a BaseLanguageModel to be
+        constructed in the future.  Neuro SAN infrastructures prefers that this
+        be an asynchronous client, however we realize some BaseLanguageModels
+        do not support that (even though they should!).
+
+        Implementations should retain any references to state that needs to be cleaned up
+        in the delete_resources() method.
+
+        :param config: The fully specified llm config
+        :return: The web client that accesses the LLM.
+                By default this is None, as many BaseLanguageModels
+                do not allow a web client to be passed in as an arg.
+        """
+        # OpenAI is the one chat class that we do not require any extra installs.
+        # This is what we want to work out of the box.
+        # Nevertheless, have it go through the same lazy-loading resolver rigamarole as the others.
+
+        # Set up a resolver to use to resolve lazy imports of classes from
+        # langchain_* packages to prevent installing the world.
+        resolver = Resolver()
+
+        # pylint: disable=invalid-name
+        AsyncAzureOpenAI = resolver.resolve_class_in_module("AsyncAzureOpenAI",
+                                                            module_name="openai",
+                                                            install_if_missing="langchain-openai")
+
+        self.create_http_client(config)
+
+        # Prepare some more complex args
+        openai_api_key: str = self.get_value_or_env(config, "openai_api_key", "AZURE_OPENAI_API_KEY")
+        if openai_api_key is None:
+            openai_api_key = self.get_value_or_env(config, "openai_api_key", "OPENAI_API_KEY")
+
+        # From lanchain_openai.chat_models.azure.py
+        default_headers: Dict[str, str] = {}
+        default_headers = config.get("default_headers", default_headers)
+        default_headers.update({
+            "User-Agent": "langchain-partner-python-azure-openai",
+        })
+
+        self.async_openai_client = AsyncAzureOpenAI(
+            azure_endpoint=self.get_value_or_env(config, "azure_endpoint",
+                                                 "AZURE_OPENAI_ENDPOINT"),
```

Comment:
Because ClientPolicy derives from EnvironmentConfiguration, we can easily use the self.get_value_or_env() method when configuring the clients too, now.